### PR TITLE
FIX-#4314: Allow passing a series of dtypes to astype.

### LINKED ISSUE #4314: Can't use astype() to assign a series of dtypes to a frame.
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: MacOS Monterey
- **Modin version** (`modin.__version__`): latest source
- **Python version**: 3.9.10
- **Code we can use to reproduce**:

```python
import modin.pandas as pd
df = pd.DataFrame([[0]])
df.astype(df.dtypes)
```

### Describe the problem
If I try to pass a series of dtypes to `astype`, I get `ValueError: The truth value of a Series is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().` However, I can do that in pandas.

### Source code / logs
<details>
<summary>Show stack trace</summary>

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Input In [10], in <module>
      1 import modin.pandas as pd
      3 df = pd.DataFrame([[0]])
----> 4 df.astype(df.dtypes)

File ~/modin/modin/pandas/base.py:921, in BasePandasDataset.astype(self, dtype, copy, errors)
    918     for column in self._query_compiler.columns:
    919         col_dtypes[column] = dtype
--> 921 new_query_compiler = self._query_compiler.astype(col_dtypes)
    922 return self._create_or_update_from_compiler(new_query_compiler, not copy)

File ~/modin/modin/core/storage_formats/pandas/query_compiler.py:1508, in PandasQueryCompiler.astype(self, col_dtypes, **kwargs)
   1507 def astype(self, col_dtypes, **kwargs):
-> 1508     return self.__constructor__(self._modin_frame.astype(col_dtypes))

File ~/modin/modin/core/dataframe/pandas/dataframe/dataframe.py:114, in lazy_metadata_decorator.<locals>.decorator.<locals>.run_f_on_minimally_updated_metadata(self, *args, **kwargs)
    112     elif apply_axis == "rows":
    113         obj._propagate_index_objs(axis=0)
--> 114 result = f(self, *args, **kwargs)
    115 if apply_axis is None and not transpose:
    116     result._deferred_index = self._deferred_index

File ~/modin/modin/core/dataframe/pandas/dataframe/dataframe.py:1003, in PandasDataframe.astype(self, col_dtypes)
   1000 except TypeError:
   1001     new_dtype = dtype
-> 1003 if dtype != np.int32 and new_dtype == np.int32:
   1004     new_dtypes[column] = np.dtype("int64")
   1005 elif dtype != np.float32 and new_dtype == np.float32:

File /usr/local/lib/python3.9/site-packages/pandas/core/generic.py:1535, in NDFrame.__nonzero__(self)
   1533 @final
   1534 def __nonzero__(self):
-> 1535     raise ValueError(
   1536         f"The truth value of a {type(self).__name__} is ambiguous. "
   1537         "Use a.empty, a.bool(), a.item(), a.any() or a.all()."
   1538     )

ValueError: The truth value of a Series is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().

```
</details>

### PR DESCRIPTION
**Pull request by mvashishtha**
Signed-off-by: mvashishtha <mahesh@ponder.io>

## What do these changes do?

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #4314
- [x] tests added and passing
- [x] module layout described at `docs/development/architecture.rst` is up-to-date 
- [x] added (Issue Number: PR title (PR Number)) and github username to release notes for next major release

### CODE REVIEW

File: modin/pandas/base.py [hunk 1]
```diff
@@ -51,6 +51,7 @@
 from .utils import is_full_grab_slice
 from modin.utils import try_cast_to_pandas, _inherit_docstrings
 from modin.error_message import ErrorMessage
+import modin.pandas as pd
```
Comment (naren-ponder) [inline modin/pandas/base.py:54]: There is another `import modin.pandas as pd` on Line 490 in this file, can we delete that and move it up here? Best practice anyhow is to have imports in the beginning.

Comment (mvashishtha) [reply modin/pandas/base.py:54]: I think I've already done that in this PR.

File: modin/pandas/base.py [hunk 2]
```diff
@@ -903,20 +904,31 @@ def asof(self, where, subset=None):
         return result
 
     def astype(self, dtype, copy=True, errors="raise"):
-        col_dtypes = {}
+        # dtype can be a series, a dict, or a scalar. If it's series or scalar,
```
[the remark this answers is not in the document]
Comment (mvashishtha) [reply modin/pandas/base.py:905]: Hm, I don't know of any type more specific than `Object` that describes the `dtype` scalar that we could use in the pytype definition.

File: modin/pandas/base.py [hunk 3]
```diff
@@ -903,20 +902,31 @@ def asof(self, where, subset=None):
...
+            if not dtype.index.is_unique:
+                raise ValueError(
+                    "The new Series of types must have a unique index, i.e. "
+                    + "it must be one-to-one mapping from column names to "
+                    + " their new dtypes."
+                )
+            dtype = {column: dtype for column, dtype in dtype.items()}
+        # If we got a series or dict originally, dtype is a dict now. Its keys
+        # must be column names.
         if isinstance(dtype, dict):
```
[the remark this answers is not in the document]
Comment (mvashishtha) [reply modin/pandas/base.py:917]: No, the `if` converts series to `dict` and I want this `if` to apply both to inputs that were originally strings and inputs that were originally series.

File: modin/pandas/base.py [hunk 4]
```diff
@@ -903,20 +902,31 @@ def asof(self, where, subset=None):
...
             if (
```
[the remark this answers is not in the document]
Comment (mvashishtha) [reply modin/pandas/base.py:918]: I think the `error` parameter is about type conversions, not about the dict/series of types itself being invalid. e.g. 

```python
import pandas as pd
df = pd.DataFrame([['a']])
# This doesn't raise an error
df.astype(int, errors='ignore')
# This raises an error
df.astype(int)
```

so it's something that we can push down to the execution layer.

File: modin/pandas/base.py [hunk 5]
```diff
@@ -903,20 +902,31 @@ def asof(self, where, subset=None):
...
                 not set(dtype.keys()).issubset(set(self._query_compiler.columns))
                 and errors == "raise"
             ):
                 raise KeyError(
-                    "Only a column name can be used for the key in"
+                    "Only a column name can be used for the key in "
                     + "a dtype mappings argument."
                 )
             col_dtypes = dtype
         else:
```
[the remark this answers is not in the document]
Comment (mvashishtha) [reply modin/pandas/base.py:927]: I think `catch_bugs_and_request_email` is meant for internal Modin errors, whereas this is an error at the level of the pandas API.

**Review by jeffreykennethli — DISMISSED** [review]

**Review by naren-ponder — DISMISSED** [review]

File: modin/pandas/test/dataframe/test_map_metadata.py [hunk 6]
```diff
@@ -483,6 +483,32 @@ def test_astype():
     with pytest.raises(KeyError):
         modin_df.astype({"not_exists": np.uint8})
 
+    # The dtypes series must have a unique index.
```
Comment (YarShev) [inline modin/pandas/test/dataframe/test_map_metadata.py:486]: If pandas has similar behavior, we should be testing both Modin and pandas via eval_general.

File: modin/pandas/test/dataframe/test_map_metadata.py [hunk 7]
```diff
@@ -483,6 +483,32 @@ def test_astype():
...
+@pytest.mark.parametrize("dtypes_are_dict", [True, False])
+def test_astype_dict_or_series_multiple_column_partitions(dtypes_are_dict):
+    # Test astype with a dtypes dict that is complex in that:
+    # - It applies to columns spanning multiple column partitions
+    # - Within a partition frame df:
+    #   - dtypes.index is not a subset of df.columns
+    #   - df.columns is not a subset of dtypes.index
+
+    pandas_df = pandas.DataFrame(test_data["int_data"])
+    modin_df = pd.DataFrame(pandas_df)
```
Comment (YarShev) [inline modin/pandas/test/dataframe/test_map_metadata.py:500]:
```suggestion
    modin_df, pandas_df = create_test_dfs(test_data["int_data"])
```

File: modin/pandas/test/dataframe/test_map_metadata.py [hunk 8]
```diff
@@ -483,6 +483,32 @@ def test_astype():
...
+    if dtypes_are_dict:
+        new_dtypes = {}
+    else:
+        new_dtypes = pandas.Series()
+    for i, column in enumerate(pandas_df.columns):
+        if i % 3 == 1:
+            new_dtypes[column] = "string"
+        elif i % 3 == 2:
+            new_dtypes[column] = float
+    df_equals(modin_df.astype(new_dtypes), pandas_df.astype(new_dtypes))
```
Comment (YarShev) [inline modin/pandas/test/dataframe/test_map_metadata.py:510]:
```suggestion
    eval_general(modin_df, pandas_df, lambda df: df.astype(new_dtypes))
```

File: modin/pandas/test/test_series.py [hunk 9]
```diff
@@ -1040,6 +1042,17 @@ def test_astype(data):
     else:
         df_equals(modin_series.astype(np.float64), pandas_result)
 
+    df_equals(
+        modin_series.astype({series_name: str}),
+        pandas_series.astype({series_name: str}),
+    )
+
+    with pytest.raises(KeyError):
+        modin_series.astype({"wrong_name": str})
```
Comment (YarShev) [inline modin/pandas/test/test_series.py:1051]: Use eval_general where possible.

**Review by YarShev — COMMENTED** [review]
Also, fix conflicts please.

File: modin/pandas/test/dataframe/test_map_metadata.py (hunk 6 above)
Comment (mvashishtha) [reply modin/pandas/test/dataframe/test_map_metadata.py:486]: Done.

File: modin/pandas/test/test_series.py (hunk 9 above)
Comment (mvashishtha) [reply modin/pandas/test/test_series.py]: Done.

**Comment by mvashishtha** [comment]
> Also, fix conflicts please.

Done.

File: modin/pandas/test/dataframe/test_map_metadata.py [hunk 10]
```diff
@@ -483,6 +483,34 @@ def test_astype():
     with pytest.raises(KeyError):
         modin_df.astype({"not_exists": np.uint8})
 
+    # The dtypes series must have a unique index.
+    eval_general(
+        modin_df,
+        expected_df,
+        lambda df: df.astype(pd.Series([str, str], index=["col1", "col1"])),
```
Comment (YarShev) [inline modin/pandas/test/dataframe/test_map_metadata.py:490]:
```suggestion
        lambda df: df.astype(
            pd.Series([str, str], index=["col1", "col1"])
            if isinstance(df, pd.DataFrame)
            else pandas.Series([str, str], index=["col1", "col1"])
        )
```
Passing a Modin object in a pandas operation can lead in undefined behavior.

**Review by YarShev — APPROVED** [review]
@mvashishtha, LGTM, thanks!

**Pull request merged**